src/components/PageDefinitions.js

```diff
@@ -176,40 +176,25 @@ class PageDefinitions extends AbstractPageDefinitions {
     return false
   }
 
-  loadFromListSpec(listSpec) {
+  async loadFromListSpec(listSpec) {
     const { dispatch, token, definitions } = this.props
-    const definitionPromises = []
-    listSpec.coordinates.forEach(component => {
-      // TODO figure a way to add these in bulk. One by one will be painful for large lists
-      const spec = EntitySpec.validateAndCreate(component)
-      if (spec) {
-        const path = spec.toPath()
-        dispatch(uiBrowseUpdateList({ add: spec }))
-        !definitions.entries[path] && definitionPromises.push(dispatch(getDefinitionsAction(token, [path])))
-      }
-    })
-
-    if (listSpec.filter) {
-      this.setState({ activeFilters: listSpec.filter })
-    }
-    if (listSpec.sortBy) {
-      this.setState({ activeSort: listSpec.sortBy })
-    }
-
-    Promise.all(definitionPromises).then(() => {
-      dispatch(
-        uiBrowseUpdateList({
-          transform: this.createTransform.call(
-            this,
-            listSpec.sortBy || this.state.activeSort,
-            listSpec.filter || this.state.activeFilters
-          )
-        })
-      )
-      if (listSpec.sortBy || listSpec.filter) {
-        this.setState({ sequence: this.state.sequence + 1 })
-      }
-    })
+    if (listSpec.filter) this.setState({ activeFilters: listSpec.filter })
+    if (listSpec.sortBy) this.setState({ activeSort: listSpec.sortBy })
+    if (listSpec.sortBy || listSpec.filter) this.setState({ sequence: this.state.sequence + 1 })
+
+    const toAdd = listSpec.coordinates.map(component => EntitySpec.validateAndCreate(component)).filter(e => e)
+    dispatch(uiBrowseUpdateList({ addAll: toAdd }))
+    const missingDefinitions = toAdd.map(spec => spec.toPath()).filter(path => !definitions.entries[path])
+    await dispatch(getDefinitionsAction(token, missingDefinitions))
+    dispatch(
+      uiBrowseUpdateList({
+        transform: this.createTransform.call(
+          this,
+          listSpec.sortBy || this.state.activeSort,
+          listSpec.filter || this.state.activeFilters
+        )
```

Comment:
This is an interesting case if the user has selected a set of sorts/filters and then drops a file that has its own set of sorts/filters. I think it makes more sense to use the sorts/filters that are in the file as opposed to the sorts/filters that the user selected. I think most users probably won't have any components or sorts/filters selected before dropping a file. I think naturally a user would go to ClearlyDefined.io and if they had a file to drop they would drop it immediately before adding any other components or sorts/filters. There is no real solution to this but just something we can keep in mind.